app/models/workflow.rb

```diff
@@ -23,7 +24,17 @@ def external_signal?
     template&.signal_setting.present?
   end
 
+  def deletable?
+    requests.any? { |request| !request.finished? } ? false : true
+  end
+
   private
+  
+  def validate_deletable
+    throw :abort unless deletable?
+
+    requests.any? { EventService.new(requests.first).workflow_deleted }
```

Comment:
We should use an after_destroy callback instead of piggybacking here in a before callback.

In the event service we may allow a null request in the constructor and pass the `workflow_id` in the argument list of `workflow_deleted` method